File: js/lib/src/ipyvis.js
let widgets = require('@jupyter-widgets/base');
let _ = require('lodash');
let object_values = require('object.values');
require('./three');
require('./ipyvis.css');
let serialization = require('./serialization');


if (!Object.values) {
  object_values.shim();
}

let {View} = require('./ViewUtils/View');

let views = new Map();

let ipyvis_version = '0.1.0';

/**
 * Getter for view
 * @param container : container of the view that you want
 * @return : the view which is in this container, undefined if there's
 * no view in this container
 * **/
function get_view (container) {
  if (views.has(container)) {
    return views.get(container);
  }

  return undefined;
}

/**
 * Resize the canvas when container is resized
 * @param container : container of the view that you want to resize
 * @param opts : optional options of your resize, if no options is given
 * the view will fill the container. Options can be for example
 * { width: 800, height: 600, backgroundColor: '#f4c842' }
 * **/
function resize_view (container, opts) {
  let view = get_view(container);

  if (view !== undefined) { view.resize(opts); }
}

/**
 * Remove a view
 * @param container : container in which you want to remove the view
 * **/
function remove_view (container) {
  let view = views.get(container);

  if (view !== undefined) {
    views.delete(container);

    return view.remove();
  }

  return false;
}

let SceneModel = widgets.DOMWidgetModel.extend({
    defaults: _.extend(widgets.DOMWidgetModel.prototype.defaults(), {
        _model_name : 'SceneModel',
        _view_name : 'SceneView',
        _model_module : 'ipyvis',
        _view_module : 'ipyvis',
        _model_module_version : ipyvis_version,
        _view_module_version : ipyvis_version,
        mesh: undefined
    })
}, {
    serializers: _.extend({
        mesh: { deserialize: widgets.unpack_models }
    }, widgets.WidgetModel.serializers)
});

let SceneView = widgets.DOMWidgetView.extend({
    render: function() {
        this.el.classList.add('ipyvis-scene');
        this.view = new View(this.el);
        views.set(this.el, this.view);

        this.model_events();

        let mesh = this.model.get('mesh');
        return this.view.addDataBlock(
            mesh.get('vertices'),
            mesh.get('faces'),
            this.get_data()
        ).then(((block) => {
            this.dataBlock = block;
            block.colored = true;
        }));
    },

    model_events: function() {
        this.model.on('change:mesh', () => {
            this.view.removeBlock(this.dataBlock);

            let mesh = this.model.get('mesh');
            return this.view.addDataBlock(
                mesh.get('vertices'),
                mesh.get('faces'),
                this.get_data()
            ).then(((block) => {
                this.dataBlock = block;
                block.colored = true;
            }));
        })
    },

    get_data: function() {
        let mesh = this.model.get('mesh');
        let mesh_data = mesh.get('data');
        let data = {};
        mesh_data.forEach((data_model) => {
            let data_name = data_model.get('name');
            data[data_name] = {};
            data_model.get('components').forEach((component_model) => {
                let component_name = component_model.get('name');
                data[data_name][component_name] = {
                    array: component_model.get('array'),
                    min: component_model.get('min'),
                    max: component_model.get('max')
                };
            });
        });

        return data;
    }
});

let ComponentModel = widgets.WidgetModel.extend({
    defaults: _.extend(widgets.WidgetModel.prototype.defaults(), {
        _model_name : 'ComponentModel',
        // _view_name : 'ComponentView',
        _model_module : 'ipyvis',
        _view_module : 'ipyvis',
        _model_module_version : ipyvis_version,
        _view_module_version : ipyvis_version,
        name: '',
        array: [],
        min: undefined,
        max: undefined
    })
}, {
    serializers: _.extend({
        array: serialization.float32array
    }, widgets.WidgetModel.serializers)
});

let DataModel = widgets.WidgetModel.extend({
    defaults: _.extend(widgets.WidgetModel.prototype.defaults(), {
        _model_name : 'DataModel',
        // _view_name : 'DataView',
        _model_module : 'ipyvis',
        _view_module : 'ipyvis',
        _model_module_version : ipyvis_version,
        _view_module_version : ipyvis_version,
        name: '',
        components: []
    })
}, {
    serializers: _.extend({
        components: { deserialize: widgets.unpack_models }
    }, widgets.WidgetModel.serializers)
});

let MeshModel = widgets.WidgetModel.extend({
    defaults: _.extend(widgets.WidgetModel.prototype.defaults(), {
        _model_name : 'MeshModel',
        // _view_name : 'MeshView',
        _model_module : 'ipyvis',
        _view_module : 'ipyvis',
        _model_module_version : ipyvis_version,
        _view_module_version : ipyvis_version,
        vertices: [],
        faces: [],
        tetras: [],
        data: []
    })
}, {
    serializers: _.extend({
        vertices: serialization.float32array,
        faces: serialization.uint32array,
        tetras: serialization.uint32array,
        data: { deserialize: widgets.unpack_models }
    }, widgets.WidgetModel.serializers)
});

module.exports = {
    SceneModel: SceneModel,
    SceneView: SceneView,
    DataModel: DataModel,
    ComponentModel: ComponentModel,
    MeshModel: MeshModel
};

File: ipyvis/ipyvis.py
from array import array

from traitlets import Unicode, List, Instance, Float
from traittypes import Array
from ipywidgets import widget_serialization, DOMWidget, Widget, register

from .serialization import array_serialization
from .vtk_loader import (
    load_file,
    get_faces, get_vertices, get_data,
    FLOAT32, UINT32
)

ipyvis_version = '^0.1.0'


@register
class Component(Widget):
    """A data component widget."""
    # _view_name = Unicode('ComponentView').tag(sync=True)
    _model_name = Unicode('ComponentModel').tag(sync=True)
    _view_module = Unicode('ipyvis').tag(sync=True)
    _model_module = Unicode('ipyvis').tag(sync=True)
    _view_module_version = Unicode(ipyvis_version).tag(sync=True)
    _model_module_version = Unicode(ipyvis_version).tag(sync=True)

    name = Unicode().tag(sync=True)
    # TODO: validate data as being 1-D array, and validate dtype
    array = Array(default_value=array(FLOAT32)).tag(sync=True, **array_serialization)
    min = Float(allow_none=True, default_value=None).tag(sync=True)
    max = Float(allow_none=True, default_value=None).tag(sync=True)


@register
class Data(Widget):
    """A data widget."""
    # _view_name = Unicode('DataView').tag(sync=True)
    _model_name = Unicode('DataModel').tag(sync=True)
    _view_module = Unicode('ipyvis').tag(sync=True)
    _model_module = Unicode('ipyvis').tag(sync=True)
    _view_module_version = Unicode(ipyvis_version).tag(sync=True)
    _model_module_version = Unicode(ipyvis_version).tag(sync=True)

    name = Unicode().tag(sync=True)
    components = List(Instance(Component)).tag(sync=True, **widget_serialization)


@register
class Mesh(Widget):
    """A 3-D Mesh widget."""
    # _view_name = Unicode('MeshView').tag(sync=True)
    _model_name = Unicode('MeshModel').tag(sync=True)
    _view_module = Unicode('ipyvis').tag(sync=True)
    _model_module = Unicode('ipyvis').tag(sync=True)
    _view_module_version = Unicode(ipyvis_version).tag(sync=True)
    _model_module_version = Unicode(ipyvis_version).tag(sync=True)

    # TODO: validate vertices/faces/tetras as being 1-D array, and validate dtype
    vertices = Array(default_value=array(FLOAT32)).tag(sync=True, **array_serialization)
    faces = Array(default_value=array(UINT32)).tag(sync=True, **array_serialization)
    tetras = Array(default_value=array(UINT32)).tag(sync=True, **array_serialization)
    data = List(Instance(Data), default_value=[]).tag(sync=True, **widget_serialization)

    @staticmethod
    def from_vtk(path):
        grid = load_file(path)

        grid_data = get_data(grid)
        data = []
        for key, value in grid_data.items():
            data.append(Data(
                name=key,
                components=[
                    Component(name=comp_name, array=comp['array'], min=comp['min'], max=comp['max'])
                    for comp_name, comp in value.items()
                ]
            ))

        return Mesh(
            vertices=get_vertices(grid),
            faces=get_faces(grid),
            tetras=array(UINT32),
            data=data
        )


@register
class Scene(DOMWidget):
    """A 3-D Scene widget."""
    _view_name = Unicode('SceneView').tag(sync=True)
    _model_name = Unicode('SceneModel').tag(sync=True)
    _view_module = Unicode('ipyvis').tag(sync=True)
    _model_module = Unicode('ipyvis').tag(sync=True)
    _view_module_version = Unicode(ipyvis_version).tag(sync=True)
    _model_module_version = Unicode(ipyvis_version).tag(sync=True)

    mesh = Instance(Mesh).tag(sync=True, **widget_serialization)
